crates/red_knot_python_semantic/src/semantic_index/usedef.rs

```diff
@@ -0,0 +1,127 @@
+use crate::ast_node_ref::AstNodeRef;
+use crate::db::Db;
+use crate::semantic_index::ast_ids::ScopedUseId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::symbol::{FileScopeId, ScopeId, ScopedSymbolId};
+use ruff_db::files::File;
+use ruff_index::IndexVec;
+use ruff_python_ast as ast;
+use salsa;
+
+/// An independently type-inferable expression.
+///
+/// Includes constraint expressions (e.g. if tests) and the RHS of an unpacking assignment.
+#[salsa::tracked]
+pub(crate) struct Expression<'db> {
+    /// The file in which the expression occurs.
+    #[id]
+    pub(crate) file: File,
+
+    /// The scope in which the expression occurs.
+    #[id]
+    pub(crate) file_scope: FileScopeId,
+
+    /// The statement node containing the expression.
+    /// We hold the statement node to clarify which specific contexts these
+    /// independently-inferable expressions can occur in, and limit the variants of this enum.
+    #[no_eq]
+    #[return_ref]
+    pub(crate) ast: ExpressionKind,
+}
+
+impl<'db> Expression<'db> {
+    /// Return the actual expression.
+    #[allow(unused)]
+    pub(crate) fn node(self, db: &'db dyn Db) -> &ast::Expr {
+        match self.ast(db) {
+            ExpressionKind::AssignedValue(node_ref) => &node_ref.node().value,
+            ExpressionKind::IfTest(node_ref) => &node_ref.node().test,
+        }
+    }
+
+    pub(crate) fn scope(self, db: &'db dyn Db) -> ScopeId<'db> {
+        self.file_scope(db).to_scope_id(db, self.file(db))
+    }
+}
+
+#[derive(Clone, Debug)]
+#[allow(dead_code)]
+pub(crate) enum ExpressionKind {
+    AssignedValue(AstNodeRef<ast::StmtAssign>),
+    IfTest(AstNodeRef<ast::StmtIf>),
+    // TODO if-expressions, loops, etc
+}
+
+/// All definitions that can reach a given use of a name.
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct UseDefMap<'db> {
+    // TODO store constraints with definitions for type narrowing
+    // TODO smallvecs?
+    /// Definitions that can reach a [`ScopedUseId`].
+    definitions_by_use: IndexVec<ScopedUseId, Vec<Definition<'db>>>,
+
+    /// Definitions of a symbol visible to other scopes.
+    public_definitions: IndexVec<ScopedSymbolId, Vec<Definition<'db>>>,
```

Comment:
Working on a change to use `Cow` here so that repeated uses with the same set of visible definitions require fewer allocations. But this can also land as a follow-up PR if I don't get it done before review of this PR.

Reply:
An alternative could be to have a separate `Vec` into which we write all these sub-vecs and instead just store a `Range` into that vec here (and uses that have the same definitions can just use the same range). But yes, I think it is worthwhile to spend some time investing if we can reduce the allocations (and clones)